packages/material-ui-unstyled/src/ButtonUnstyled/useButton.ts

```diff
@@ -0,0 +1,226 @@
+    }
+
+    otherHandler();
+  };
+
+  const dispatchClickEvent = (preventDefault: boolean) => {
+    const clickEvent = new window.MouseEvent('click', {
+      bubbles: true,
+      cancelable: true,
+      view: window,
+    });
+
+    if (preventDefault) {
+      clickEvent.preventDefault();
+    }
+
+    buttonRef.current?.dispatchEvent(clickEvent);
+  };
+
+  const handleFocus = useEventCallback(
+    (event: React.FocusEvent<HTMLButtonElement>, otherHandler: OtherEventHandler) => {
+      // Fix for https://github.com/facebook/react/issues/7769
+      if (!buttonRef.current) {
+        buttonRef.current = event.currentTarget;
+      }
+
+      handleFocusVisible(event);
+      if (isFocusVisibleRef.current === true) {
+        setFocusVisible(true);
+        onFocusVisible?.(event);
+      }
+
+      otherHandler();
+    },
+  );
+
+  const elementType = component ?? components.Root ?? 'button';
+
+  const isNonNativeButton = () => {
+    const button = buttonRef.current;
+    return (
+      elementType !== 'button' && !(button?.tagName === 'A' && (button as HTMLAnchorElement)?.href)
+    );
+  };
+
+  const handleMouseDown = (
+    event: React.MouseEvent<HTMLButtonElement>,
+    otherHandler: OtherEventHandler,
+  ) => {
+    if (event.target === event.currentTarget && !disabled) {
+      setActive(true);
+    }
+
+    otherHandler();
+  };
+
+  const handleMouseUp = (
+    event: React.MouseEvent<HTMLButtonElement>,
+    otherHandler: OtherEventHandler,
+  ) => {
+    if (event.target === event.currentTarget) {
+      setActive(false);
+    }
+
+    otherHandler();
+  };
+
+  const handleKeyDown = useEventCallback(
+    (event: React.KeyboardEvent, otherHandler: OtherEventHandler) => {
+      if (event.target === event.currentTarget && isNonNativeButton() && event.key === ' ') {
+        event.preventDefault();
+      }
+
+      if (event.target === event.currentTarget && event.key === ' ' && !disabled) {
+        setActive(true);
+      }
+
+      otherHandler();
+
+      // Keyboard accessibility for non interactive elements
+      if (
+        event.target === event.currentTarget &&
+        isNonNativeButton() &&
+        event.key === 'Enter' &&
+        !disabled
+      ) {
+        event.preventDefault();
+        dispatchClickEvent(true);
+      }
+    },
+  );
+
+  const handleKeyUp = useEventCallback(
+    (event: React.KeyboardEvent, otherHandler: OtherEventHandler) => {
+      // calling preventDefault in keyUp on a <button> will not dispatch a click event if Space is pressed
+      // https://codesandbox.io/s/button-keyup-preventdefault-dn7f0
+
+      if (event.target === event.currentTarget) {
+        setActive(false);
+      }
+
+      otherHandler();
+
+      // Keyboard accessibility for non interactive elements
+      if (
+        event.target === event.currentTarget &&
+        isNonNativeButton() &&
+        event.key === ' ' &&
+        !event.defaultPrevented
+      ) {
+        dispatchClickEvent(false);
```

Comment:
It's just not how we ever implemented custom event handlers. So when I say "unforseen" I really meen "unforseen". If I had a concrete example, it wouldn't be unforseen. At least as far as I translate "unforseen".

I can give you an example for synthetic events in general: https://github.com/mui-org/material-ui/pull/22524 had to be reverted because it caused https://github.com/mui-org/material-ui/issues/27664.